enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,336 @@
+
+const (
+	ActivePhase LocalVolumeGroupPhase = "Active"
+	FailedPhase LocalVolumeGroupPhase = "Failed"
+)
+
+type LocalVolumeGroupStatus struct {
+	// Phase describes the state of the LocalVolumeGroup
+	Phase LocalVolumeGroupPhase `json:"phase,omitempty"`
+
+	// A human-readable message indicating details about why the LocalVolumeGroup is in this state.
+	// +optional
+	Message string `json:"message,omitempty"`
+
+	// Reason is a brief CamelCase string that describes any failure and is meant
+	// for machine parsing and tidy display in the CLI.
+	// +optional
+	Reason string `json:"reason,omitempty"`
+
+	// TotalProvisionedDeviceCount is the count of the total devices over which the PVs has been provisioned
+	TotalProvisionedDeviceCount *int32 `json:"totalProvisionedDeviceCount,omitempty"`
+
+	// observedGeneration is the last generation change the operator has dealt with
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+}
+```
+### Workflow of LocalPV creation via LocalVolumeGroup CR
+- Once the `LocalVolumeGroup` is created, the controller will:
+  - configure the local-static-provisioner to make a new StorageClass based on certain directories on the selected nodes.
+  - assign diskmaker daemons to the selected nodes.
+- The diskmaker daemon will find devices that match the disovery policy and symlink them into the directory that the local-static-provisioner is watching.
+- Possible `status.phase` values for `LocalVolumeGroup`
+  - `Active`
+  - `Failed`
+
+#### Note: There is a chance of race condition: 
+- If two `LocalVolumeGroup` CR targets same nodes with overlapping inclusion filter.
+- If `LocalVolumeGroup` CR and `LocalVolume` CR targets the same node with overlapping devices.
+  
+Example of `LocalVolumeGroup` CR:
+
+```yaml
+apiVersion: local.storage.openshift.io/v1alpha1
+kind: LocalVolumeGroup
+metadata:
+  name: example-autodetect
+spec:
+  nodeSelector:
+    nodeSelectorTerms:
+      - matchExpressions:
+          - key: kubernetes.io/hostname
+            operator: In
+            values:
+              - worker-0
+              - worker-1
+  storageClassName: example-storageclass
+  volumeMode: Block
+  minDeviceCount: 5
+  maxDeviceCount: 10
+  deviceInclusionSpec:
+    deviceTypes:
+      - disk
+    deviceMechanicalProperty: Rotational
+    minSize: 10G
+    maxSize: 100G
+status:
+  phase: Active
+  totalProvisionedDeviceCount: 4
+```
+
+```yaml
+apiVersion: local.storage.openshift.io/v1alpha1
+kind: LocalVolumeGroup
+metadata:
+  name: example-autodetect
+spec:
+  nodeSelector:
+    nodeSelectorTerms:
+      - matchExpressions:
+          - key: kubernetes.io/hostname
+            operator: In
+            values:
+              - worker-0
+              - worker-1
+  storageClassName: example-storageclass
+  volumeMode: filesystem
+  fstype: ext4
+  minDeviceCount: 5
+  maxDeviceCount: 10
+  deviceInclusionSpec:
+    deviceTypes:
+      - disk
+      - nvme
```

Comment:
I thought `disk` and `partition` were the possible values here.

Reply:
yeah

Reply:
changed